hazelcast/src/main/java/com/hazelcast/instance/impl/JetExtension.java

```diff
@@ -111,17 +119,31 @@ public void beforeClusterStateChange(ClusterState requestedState) {
     }
 
     public void onClusterStateChange(ClusterState ignored) {
-        jetService.getJobCoordinationService().clusterChangeDone();
+        if (isActivated()) {
+            jetService.getJobCoordinationService().clusterChangeDone();
+        }
+    }
+
+    public void onClusterVersionChange(Version newVersion) {
+        if (!isActivated() && newVersion.isGreaterOrEqual(Versions.V5_0)) {
+            setActivated(true);
+            jetService.getJobCoordinationService().startScanningForJobs();
+            logger.info("Jet extension is enabled after the cluster version upgrade.");
+        }
     }
 
     public void beforeShutdown(boolean terminate) {
-        if (!terminate) {
+        if (!terminate && isActivated()) {
             jetService.shutDownJobs();
         }
     }
 
     public void handlePacket(Packet packet) {
-        jetService.handlePacket(packet);
+        if (isActivated()) {
```

Comment:
Instead of doing a volatile read for each incoming packet and impose a penalty on the happy path, I 'd rather leave this method unchanged. If cluster version is < 5.0 and a Jet packet is received, then let it fail further down its processing.